File: lab1/CoolFilms/CoolFilms/ViewModel/RegistrationViewModel.swift
import FirebaseAuth
import FirebaseDatabase
import Combine

class RegistrationViewModel: ObservableObject {
    @Published var login = ""
    @Published var email = ""
    @Published var password = ""
    @Published var alert = ""
    @Published var isShowFilmsListView = false
    @Published var isShowAlert = false
    
    private let auth = Auth.auth()
    private let database = Database.database().reference()
    
    func hidePopup(){
        isShowAlert = false
    }
    
    func showAlert(message: String) {
        alert = message
        isShowAlert = true
    }
    
    func registerUser() {
        guard login.count >= 4, password.count >= 8, !email.isEmpty else {
            showAlert(message: "Invalid data in input fields")
            return
        }
        
        auth.createUser(withEmail: email, password: password) { [weak self] result, error in
            guard let self = self else { return }
            
            if let _ = error {
                self.showAlert(message: "Error creating user")
                return
            }
            
            guard let user = result?.user else {
                self.showAlert(message: "No user created")
                return
            }
            
            let userData: [String: Any] = [
                "login": self.login,
                "email": self.email
            ]
            
            self.database.child("users").child(user.uid).setValue(userData) { error, _ in
                if error != nil {
                    self.showAlert(message: "Error saving user data")
                    return
                }
                
                self.isShowFilmsListView = true
            }
        }
    }
}

File: lab3/lib/viewmodel/registration_viewmodel.dart
import 'package:firebase_auth/firebase_auth.dart';
import 'package:firebase_database/firebase_database.dart';
import 'package:rxdart/rxdart.dart';

class RegistrationViewModel {
  final BehaviorSubject<String> login = BehaviorSubject<String>.seeded('');
  final BehaviorSubject<String> email = BehaviorSubject<String>.seeded('');
  final BehaviorSubject<String> password = BehaviorSubject<String>.seeded('');
  final BehaviorSubject<String> alert = BehaviorSubject<String>.seeded('');
  final BehaviorSubject<bool> isShowFilmsListView = BehaviorSubject<bool>.seeded(false);
  final BehaviorSubject<bool> isShowAlert = BehaviorSubject<bool>.seeded(false);

  final FirebaseAuth auth = FirebaseAuth.instance;
  final FirebaseDatabase database = FirebaseDatabase.instance;

  void hidePopup() {
    isShowAlert.add(false);
  }

  void showAlert(String message) {
    alert.add(message);
    isShowAlert.add(true);
  }

  void registerUser() async {
    if (login.value.length < 4 || password.value.length < 8 || email.value.isEmpty) {
      showAlert("Invalid data in input fields");
      return;
    }

    try {
      UserCredential userCredential = await auth.createUserWithEmailAndPassword(
        email: email.value,
        password: password.value,
      );

      User? user = userCredential.user;

      if (user == null) {
        showAlert("No user created");
        return;
      }

      Map<String, dynamic> userData = {
        "login": login.value,
        "email": email.value,
      };

      await database.ref("users").child(user.uid).set(userData);

      isShowFilmsListView.add(true);
    } on FirebaseAuthException catch (e) {
      showAlert("Error creating user: ${e.message}");
    } catch (e) {
      showAlert("Error creating user: $e");
    }
  }

  void dispose() {
    login.close();
    email.close();
    password.close();
    alert.close();
    isShowFilmsListView.close();
    isShowAlert.close();
  }
}